native/membrane_videocompositor_opengl_rust/src/errors.rs

```diff
@@ -0,0 +1,64 @@
+use crate::gl;
+use thiserror::Error;
+
+#[derive(Debug, rustler::NifStruct)]
+#[module = "Membrane.VideoCompositor.OpenGL.Rust.ErrorLocation"]
+/// Carries information about a location where an error happened
+pub struct ErrorLocation {
+    pub file: String,
+    pub line: u32,
+    pub call: String,
+}
+
+pub fn result_or_gl_error<T>(
+    res: T,
+    err: u32,
+    location: ErrorLocation,
+) -> Result<T, CompositorError> {
+    match err {
+        gl!(NO_ERROR) => Ok(res),
+        gl!(INVALID_ENUM) => Err(CompositorError::GLError("gl_invalid_enum", location)),
+        gl!(INVALID_VALUE) => Err(CompositorError::GLError("gl_invalid_value", location)),
+        gl!(INVALID_OPERATION) => Err(CompositorError::GLError("gl_invalid_opegation", location)),
+        gl!(INVALID_FRAMEBUFFER_OPERATION) => Err(CompositorError::GLError(
+            "gl_invalid_framebuffer_operations",
+            location,
+        )),
+        gl!(OUT_OF_MEMORY) => Err(CompositorError::GLError("gl_out_of_memory", location)),
```

Comment:
what does these `gl!` expand into?

Reply:
That's the second case in the macro
```rs
($name:ident) => {
        glad_gles2::gl::$name
    }
```

Reply:
so it expands to `glad_gles2::gl::INVALID_ENUM`

Reply:
ok, gotcha; though this is quite not obvious and I would rather avoid such shortcuts in code